_posts/2018-10-23-artsy-auth.md

```diff
@@ -0,0 +1,279 @@
+that Gravity is going to use to encrypt our JWT, so we need to make sure that our app has access to these two
+values.
+
+Go back to your app, open `_config.rb` and add the following:
+
+```
+  config.app_id=ENV['APP_ID']
+  config.app_secret=ENV['APP_SECRET']
+  config.gravity_url=ENV['GRAVITY_URL']
+```
+
+Then in your `.env` file (or however you're managing environment variables) go ahead and set
+`APP_ID=02a971b9401a317bf815`, `APP_SECRET=8b7d9b597428254f9305544b8cd4d710` and
+`GRAVITY_URL=https://staging-api.artsy.net`. These values can now be accessed from `yourapp.config.app_id`,
+`yourapp.config.app_secret` and `yourapp.config.gravity_url` - we're going to need these values to configure
+ArtsyAuth in our new application.
+
+This is enough, then, to enable user authentication! If we want to create application authentication we need to
+take one more step.
+
+#### Create a JWT for Application Trust
+
+For users, JWTs are generated when they log in. But if we are trying to send HTTP requests from barapp to fooapp
+we've established that no login is going to take place - so how does barapp get the JWT it needs to authenticate
+against our fooapp? We're going to create it manually and store it in barapp's environment. Let's go back to our
+Gravity console session - make sure `ca` is still defined. If you closed the session then before running the next
+code you should get a reference to that client application. `ca = ClientApplication.where(name: 'fooapp').first`
+should do it. Then:
+
+```
+  gravity:staging> expires_in = 10.years.from_now
+  gravity:staging> token = ApplicationTrust.create_for_token_authentication(ca, expires_in: expires_in)
+```
+
+This should generate a long string that it dumps to the console. Congratulations - that string is an encrypted JWT.
+Any http request with the header `Authorization: Bearer <encrypted>` will now be accepted by fooapp.
+
+So to use this JWT, we copy the string into barapp's environment - often with a second environment variable for the
+specific host we want to connect to. So in barapp you'd have environment variables like `FOO_JWT=<string>` and
+`FOO_HOST=http://fooapp.artsy.net` or `FOO_HOST=http://fooapp-staging.artsy.net` or whatever. (We store the host as
+an environment variable rather than having it in the code precisely so that we can target specific instances of the
+application depending on the desired context).
+
+Somewhere in barapp we'll have a service that sends HTTP requests to fooapp - we just have to make sure that those
+requests have the authorization header with the correct encrypted token, and bam! We're authenticated!
+
+So now it's time to put it all together - we know how to generate tokens, we know how to attach them to requests as
+authorization headers, and we know how to decrypt them locally in fooapp. So now what? What do we do with those
+tokens once we have them? That's where a useful library called ArtsyAuth comes in, at least if you're writing a
```

Comment:
Worth a link to ArtsyAuth on the first time you mention it - https://rubygems.org/gems/artsy-auth/versions/0.1.3